File: server/src/main/kotlin/com/evgenltd/financemanager/importexport/service/CategoryMappingService.kt
package com.evgenltd.financemanager.importexport.service

import com.evgenltd.financemanager.common.repository.and
import com.evgenltd.financemanager.common.repository.eq
import com.evgenltd.financemanager.common.repository.find
import com.evgenltd.financemanager.common.repository.findAllByCondition
import com.evgenltd.financemanager.importexport.converter.CategoryMappingConverter
import com.evgenltd.financemanager.importexport.entity.CategoryMapping
import com.evgenltd.financemanager.importexport.record.CategoryMappingFilter
import com.evgenltd.financemanager.importexport.record.CategoryMappingPage
import com.evgenltd.financemanager.importexport.record.CategoryMappingRecord
import com.evgenltd.financemanager.importexport.repository.CategoryMappingRepository
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class CategoryMappingService(
    private val categoryMappingRepository: CategoryMappingRepository,
    private val categoryMappingConverter: CategoryMappingConverter
) {

    fun list(filter: CategoryMappingFilter): CategoryMappingPage =
        categoryMappingRepository.findAllByCondition(filter.page, filter.size) {
            (CategoryMapping.Companion::parser eq filter.parser?.id) and
            (CategoryMapping.Companion::categoryId eq filter.category?.id)
        }.let {
            CategoryMappingPage(
                total = it.totalElements,
                page = filter.page,
                size = filter.size,
                mappings = it.content
                    .sortedBy { it.pattern }
                    .map { entry -> categoryMappingConverter.toRecord(entry) }
            )
        }

    fun byId(id: UUID): CategoryMappingRecord = categoryMappingRepository.find(id).let { categoryMappingConverter.toRecord(it) }

    fun update(record: CategoryMappingRecord) {
        categoryMappingRepository.save(categoryMappingConverter.toEntity(record))
    }

    fun delete(id: UUID) {
        categoryMappingRepository.deleteById(id)
    }

    fun findByParser(parser: UUID): List<Pair<Regex, CategoryMapping>> =
        categoryMappingRepository.findByParser(parser).map { ".*${it.pattern}.*".toRegex() to it }

}

File: web/src/app/importexport/service/category-mapping.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {CategoryMapping, CategoryMappingPage} from "../model/category-mapping";

@Injectable({
  providedIn: 'root'
})
export class CategoryMappingService {

  constructor(private http: HttpClient) {}

  list(filter: any): Observable<CategoryMappingPage> {
    return this.http.post<CategoryMappingPage>('/category-mapping/filter', filter)
  }

  byId(id: string): Observable<CategoryMapping> {
    return this.http.get<CategoryMapping>('/category-mapping/' + id)
  }

  update(categoryMapping: CategoryMapping): Observable<void> {
    return this.http.post<void>('/category-mapping', categoryMapping)
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>('/category-mapping/' + id)
  }

}
